dracut/40clevis-unlock/clevis-unlock.service

```diff
@@ -0,0 +1,23 @@
+[Unit]
+Description=Unlock disks using Clevis
+DefaultDependencies=no
+# clevis-luks-generic-unlocker depends on executables in /sysusr/usr.
+RequiresMountsFor=/sysusr/usr/
+# dracut-initqueue will wait for /dev/disk/by-label/ROOT, so we must decrypt the root disk before dracut-initqueue.
+Before=dracut-initqueue.service
+Wants=systemd-networkd.service
```

Comment:
Clevis itself doesn't really require networking, or? Only when used with tang. Having networking in every initrd boot slows things down and it also may not work with a static network setup.

Reply:
Could we use a flag file or cmdline option to opt-into clevis?

Reply:
We could add `ConditionKernelCommandLine=|flatcar.clevis_initrd_unlock=1` to the `clevis-unlock.service` unit file and require the user to ensure that this command-line option is set if needed, by either setting it via Ignition `kernelArguments` or in their PXE boot config. I think this should work, but, from the user perspective, it'd be nicer if configuring disk encryption in Ignition just worked, without needing to remember to change some other bit of configuration somewhere else.

Another solution that doesn't require the user to set boot args would be to change `clevis-unlock.service` as follows: 1) remove `Require=systemd-networkd`, 2) do not let it run the unlocker script directly, but instead let it run a script that first checks whether there are any disks to decrypt (e.g., via lsblk); if not, let it terminate, and if so, let it ensure systemd-networkd is started and then call the unlocker script. This is a little more complicated for us than the kernel command line option but maybe better from the user perspective.

My hunch is that the second solution is preferable (if it's possible to get this to work). What do you think---shall I give it a shot?

Reply:
I would say a commandline that requests networking is ok, but we should support clevis without networking too (only tpm).

Look at the clevis repo, there are examples there that iterate over luks devices and check whether any of them have a tang pin. We can have a generator do that and then pull in systemd-networkd as required.

Reply:
Good, so we want to be able to 1) not run clevis in the initramfs at all, 2) run clevis without network (tpm only), 3) run clevis with network. I implemented this as follows

- `rd.clevis_unlock=network` runs clevis with network dependencies
- `rd.clevis_unlock=no_network` runs clevis without network dependencies (TPM only)
- If neither is set, then clevis doesn't run at all.

Happy to change the name of the command line parameter if you'd prefer different names.

@jepio Good idea! For now, I used command line parameters to control in the generator whether systemd-networkd is pulled in, but a generator that looks for Tang-encrypted disks would work as well. Slight complication: generators are run before `sysusr-usr.mount` and we currently overwrite the clevis dependencies with [symlinks]((https://github.com/flatcar/bootengine/pull/83#discussion_r1448032002)) in the initramfs (maybe to keep the initramfs image small, @pothos ?). If we want to use a generator that reads the LUKS headers, we probably need to have some of the actual clevis files rather than symlinks in the initramfs image.

After adding the necessary dependencies, changing the generator to something like this should do the job:

```bash
#!/bin/bash

set -e

UNIT_DIR="$1"

function exists_tang_encrypted_disk() {
    local _d 
    for _d in $(lsblk -o PATH,FSTYPE,RM \
               | awk '$2 == "crypto_LUKS" && $3 == "0" { print $1 }');
    do
        if bindings="$(clevis-luks-list -d "${_d}" 2>/dev/null)" \
                         && [[ "${bindings}" == *tang* ]]; then
            return 0
        fi
    done

    return 1
}

function add_clevis_network_dependencies() {
    mkdir -p "${UNIT_DIR}/clevis_unlock.service.d"

    cat >${UNIT_DIR}/clevis_unlock.service.d/networkd_dependency.conf <<EOF
    # Automatically generated by tang-network-generator

    [Unit]
    Wants=systemd-networkd.service systemd-resolved.service network-online.target
    After=systemd-networkd.service systemd-resolved.service network-online.target
EOF
}

if exists_tang_encrypted_disk ; then
    add_clevis_network_dependencies
fi
```

Since the user now has to explicitly enable clevis unlock anyways, having the user specify whether they need network or not seems not so bad from a UX perspective, and maybe it's not worth the extra complexity to detect this automatically at this point. I don't have a strong opinion here and would both be happy to stick with the non-automated, current solution, or to change it to the automated one.

Reply:
It could also be a helper script in `ExecStartPre` that starts networking when needed (`systemctl start --quiet systemd-networkd systemd-resolved`).

I see it as an optimization for a follow-up and we can focus on getting it merged as is, first.